File: internal/platform/observability/otel.go
package observability

import (
	"context"
	"errors"
	"net/http"
	"os"
	"strings"

	"go.opentelemetry.io/contrib/exporters/autoexport"
	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// SetupOTel installs a real OpenTelemetry tracer provider ONLY when the operator
// configured an exporter (OTEL_TRACES_EXPORTER or OTEL_EXPORTER_OTLP_ENDPOINT
// set). Unconfigured ⇒ no SDK install, the global tracer stays the no-op default,
// and the otel chain link is skipped — zero overhead, byte-identical boot. This
// env-gated, inert-by-default behaviour is REQUIRED per Wave Z spec Z-1: the
// single-host default deployment (D-1) ships no collector, so OTel must cost
// nothing until a second host or an external-tenant SLA turns it on (REQ-OBS-3).
//
// The W3C TraceContext + Baggage composite propagator is installed so inbound
// `traceparent` headers are honoured and outbound calls carry the active trace
// (REQ-OBS-2). The existing X-Trace-Id path keeps working: httpObs bridges to the
// active span's TraceID when one is present (see Wrap), and falls back to
// requesttrace otherwise.
//
// serviceName sets the resource service.name attribute; if the OTEL_SERVICE_NAME
// environment variable is set (trimmed, non-empty) it overrides serviceName.
//
// Returns a shutdown func (always non-nil — a no-op when disabled) and whether
// the SDK was installed. autoexport reads OTEL_TRACES_EXPORTER (otlp|console|
// none); "console" writes spans to stdout.
func SetupOTel(ctx context.Context, serviceName string) (shutdown func(context.Context) error, enabled bool, err error) {
	noop := func(context.Context) error { return nil }

	// OTEL_TRACES_EXPORTER=none (the OTel-spec "explicitly off" value, matched
	// case-insensitively + trimmed) short-circuits before any SDK install —
	// return the no-op provider even when an OTLP endpoint is also set. The
	// empty-exporter + empty-endpoint case stays off too (unconfigured default).
	exporter := strings.ToLower(strings.TrimSpace(os.Getenv("OTEL_TRACES_EXPORTER")))
	if exporter == "none" {
		return noop, false, nil
	}
	if exporter == "" && os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT") == "" {
		return noop, false, nil
	}

	exp, err := autoexport.NewSpanExporter(ctx)
	if err != nil {
		return noop, false, err
	}

	if envName := strings.TrimSpace(os.Getenv("OTEL_SERVICE_NAME")); envName != "" {
		serviceName = envName
	}
	res, err := resource.Merge(
		resource.Default(),
		resource.NewWithAttributes("", attribute.String("service.name", serviceName)),
	)
	if err != nil && !errors.Is(err, resource.ErrSchemaURLConflict) {
		return noop, false, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, true, nil
}

// OTelMiddleware returns the chain link that wraps the handler in an
// otelhttp.Handler, naming each server span by a low-cardinality route. It is
// only invoked when SetupOTel reported enabled=true; when OTel is off, main.go
// passes nil for this link and buildChain skips it (zero overhead).
//
// Span name resolution mirrors httpObs's routeLabel: prefer r.Pattern (Go 1.22
// ServeMux), which otelhttp re-evaluates after the inner mux matches. Auth/iam
// middlewares between this link and the mux re-create the request via
// WithContext, so r.Pattern often does not propagate back to the instance
// otelhttp holds; in that case fall back to normalizeRoute(path) — the same
// hand-maintained collapse used for metrics — instead of the raw URL, keeping
// span cardinality bounded regardless (REQ-OBS-1).
func OTelMiddleware() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return otelhttp.NewHandler(next, "http.server",
			otelhttp.WithSpanNameFormatter(func(_ string, r *http.Request) string {
				if r.Pattern != "" {
					if _, p, ok := strings.Cut(r.Pattern, " "); ok && p != "" {
						return r.Method + " " + p
					}
					return r.Pattern
				}
				return r.Method + " " + normalizeRoute(strings.ReplaceAll(r.URL.Path, "\n", ""))
			}),
		)
	}
}

File: apps/docx-renderer/src/observability/otel.ts
// Mirrors internal/platform/observability/otel.go's SetupOTel contract so the
// 4th binary (docx-renderer, Node/TS) follows the SAME env conventions as the
// 3 Go binaries: unconfigured (no OTEL_TRACES_EXPORTER and no
// OTEL_EXPORTER_OTLP_ENDPOINT) installs nothing and the global tracer stays
// the OTel no-op default — zero overhead, no log spam, no crash. Configured
// (OTEL_TRACES_EXPORTER=otlp|console, or OTEL_EXPORTER_OTLP_ENDPOINT set)
// installs a real NodeTracerProvider exporting via OTLP/HTTP. This closes
// T-010 (wiki/modules/editor-ui-eigenpal-tech-debt.md) — REQ-OBS-3's last
// unpropagated hop (worker → docx-renderer).
import { trace, type Tracer } from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { ConsoleSpanExporter, SimpleSpanProcessor, BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions';

const DEFAULT_SERVICE_NAME = 'docx-renderer';

export interface OTelHandle {
  readonly enabled: boolean;
  readonly propagator: W3CTraceContextPropagator;
  shutdown(): Promise<void>;
}

const noopPropagator = new W3CTraceContextPropagator();

function noopHandle(): OTelHandle {
  return {
    enabled: false,
    propagator: noopPropagator,
    shutdown: async () => {},
  };
}

/**
 * Installs a NodeTracerProvider ONLY when the operator configured an exporter
 * (OTEL_TRACES_EXPORTER or OTEL_EXPORTER_OTLP_ENDPOINT set) — same gate as the
 * Go SetupOTel. Unconfigured => returns a disabled no-op handle; the global
 * @opentelemetry/api tracer stays the built-in no-op, so span creation calls
 * anywhere in the renderer cost nothing and never throw.
 *
 * serviceName defaults to "docx-renderer"; OTEL_SERVICE_NAME overrides it,
 * matching the Go convention (SetupOTel's serviceName + OTEL_SERVICE_NAME).
 */
export function setupOTel(serviceName: string = DEFAULT_SERVICE_NAME): OTelHandle {
  const exporterEnv = (process.env.OTEL_TRACES_EXPORTER ?? '').trim().toLowerCase();
  const endpoint = (process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? '').trim();

  // OTEL_TRACES_EXPORTER=none is the OTel-spec "explicitly off" value; matches
  // Go's short-circuit even when an endpoint is also set.
  if (exporterEnv === 'none') {
    return noopHandle();
  }
  if (exporterEnv === '' && endpoint === '') {
    return noopHandle();
  }

  const resolvedServiceName = (process.env.OTEL_SERVICE_NAME ?? '').trim() || serviceName;

  const exporter =
    exporterEnv === 'console'
      ? new ConsoleSpanExporter()
      : new OTLPTraceExporter(endpoint ? { url: `${endpoint.replace(/\/$/, '')}/v1/traces` } : {});

  // Console exporter should flush synchronously (SimpleSpanProcessor) so a
  // dev/test run sees spans immediately, mirroring the Go autoexport
  // console path's directness; OTLP uses BatchSpanProcessor for real export.
  const processor =
    exporterEnv === 'console' ? new SimpleSpanProcessor(exporter) : new BatchSpanProcessor(exporter);

  const provider = new NodeTracerProvider({
    resource: resourceFromAttributes({ [ATTR_SERVICE_NAME]: resolvedServiceName }),
    spanProcessors: [processor],
  });
  provider.register({ propagator: new W3CTraceContextPropagator() });

  return {
    enabled: true,
    propagator: new W3CTraceContextPropagator(),
    shutdown: () => provider.shutdown(),
  };
}

/** The renderer's tracer — a thin wrapper so callers don't import @opentelemetry/api directly. */
export function getTracer(): Tracer {
  return trace.getTracer('docx-renderer');
}
